quota/quota.go

```diff
@@ -0,0 +1,315 @@
+		totalAvailable := quantity.Size(0)
+		for _, child := range grp.parentGroup.subGroups {
+			totalAvailable += child.MemoryLimit
+		}
+		if totalAvailable+grp.MemoryLimit > grp.parentGroup.MemoryLimit {
+			remaining := grp.parentGroup.MemoryLimit - totalAvailable
+			return fmt.Errorf("sub-group memory limit of %s is too large to fit inside remaining quota space %s for parent group %s", grp.MemoryLimit.IECString(), remaining.IECString(), grp.parentGroup.Name)
+		}
+	}
+
+	return nil
+}
+
+// NewSubGroup creates a new sub group under the current group
+func (grp *Group) NewSubGroup(name string, memLimit quantity.Size) (*Group, error) {
+	subGrp := &Group{
+		Name:        name,
+		MemoryLimit: memLimit,
+		ParentGroup: grp.Name,
+		parentGroup: grp,
+	}
+
+	if err := subGrp.validate(); err != nil {
+		return nil, err
+	}
+
+	// also double check that the sub group name is not the same as that of the
+	// parent, this is fine in systemd world, but in snapd we want unique quota
+	// groups
+	if name == grp.Name {
+		return nil, fmt.Errorf("cannot use same name %q for sub group as parent group", name)
+	}
+
+	// save the details of this new sub-group in the parent group
+	grp.subGroups = append(grp.subGroups, subGrp)
+	grp.SubGroups = append(grp.SubGroups, name)
+
+	return subGrp, nil
+}
+
+func NewGroup(name string, memLimit quantity.Size) (*Group, error) {
+	grp := &Group{
+		Name:        name,
+		MemoryLimit: memLimit,
+	}
+
+	if err := grp.validate(); err != nil {
+		return nil, err
+	}
+
+	return grp, nil
+}
+
+// CompleteCrossReferences takes a set of deserialized groups and sets all
+// cross references amongst them using the unexported fields which are not
+// serialized.
+func CompleteCrossReferences(grps map[string]*Group) error {
+	// iterate over all groups, looking for sub-groups which need to be threaded
+	// together with their respective parent groups from the set
+	for _, grp := range grps {
+		// first thread the parent link
+		if grp.ParentGroup != "" {
+			parent, ok := grps[grp.ParentGroup]
+			if !ok {
+				return fmt.Errorf("internal error: missing group %q referenced as the parent of group %q", grp.ParentGroup, grp.Name)
+			}
+			grp.parentGroup = parent
+		}
+
+		// now thread any child links from this group to any children
+		if len(grp.SubGroups) != 0 {
+			grp.subGroups = make([]*Group, len(grp.SubGroups))
+			for i, subName := range grp.SubGroups {
+				sub, ok := grps[subName]
+				if !ok {
+					return fmt.Errorf("internal error: missing group %q referenced as the child of group %q", subName, grp.Name)
+				}
+				grp.subGroups[i] = sub
+			}
+		}
+	}
+
+	return nil
+}
+
+// subTree recursively returns all of the sub-groups of the group
+func (grp *Group) subTree() []*Group {
+	subTreeList := grp.subGroups
+	for _, sub := range grp.subGroups {
+		subTreeList = append(subTreeList, sub.subTree()...)
+	}
+
+	return subTreeList
+}
+
+func deleteGrpFromList(grps []*Group, grpToDelete *Group) []*Group {
```

Comment:
givent that is is functional I suppose this should be called exlude... maybe ? (I know it would likely be called delete in a language where things are all functional)

Reply:
this helper actually goes away with your proposed implementation from the other PR